### ISSUE TITLE
Method to copy annotation objects from one page on one document onto another page in another document

### ISSUE THREAD

**Issue by amole-arup** [pymupdf/PyMuPDF#1310]
**Is your feature request related to a problem? Please describe.**
My problem is that I want to collate annotations (markups) from one PDF onto another. These would be PDFs handled by the same software, so there should be compatibility.

**Describe the solution you'd like**
 I would like a method for adding annotations to a page after extracting them from another page.

**Describe alternatives you've considered**
I could identify the type of the annotations that I have extracted and could then create them. Unfortunately, this would require a lot of programming that would seem to be unnecessary (and difficult for me and others with less programming skills).

**Additional context**
NA

**Comment by JorjMcKie** [comment]
Making sure I understand you well:
To copy pages from another PDF into the current one, you can use `doc.insert_pdf()`. This method offers to include annotations alongside the page itself.
This is not what you mean - right?

Other solutions would require some programming, as you say. But not wouldn't be that bad presumably. What type are these annotations? FreeText, Stamps, Rectangles?

**Comment by amole-arup** [comment]
Hi Jorj, thank you for your response. I have identical PDF pages in separate documents, each with annotations. I want to have one page and to copy all the annotations onto that one page in one document.

These are clouds with attached text (in red in the image). I have created a sample document with original annotations and have created a new polygon based on the same vertices as the original "not flattened" clouded note (which is also shown in red), but this is just straight lines and it does not have the attached text.

![image](https://user-images.githubusercontent.com/77268469/135986806-cd04ade1-2b65-42be-a6c8-491fafa37400.png)
[PDF sample (new).pdf](https://github.com/pymupdf/PyMuPDF/files/7284070/PDF.sample.new.pdf)

**Comment by amole-arup** [comment]
Perhaps this image will explain the process more clearly:

![image](https://user-images.githubusercontent.com/77268469/135991403-c3183a71-df66-4c18-b365-758db28121a1.png)

**Comment by JorjMcKie** [comment]
Good picture, now I understand!
Presumably, the sets of annotation on each of the identical pages are also not disjoint, right? Meaning there may be an annotation on page 1, which is also present on page 2?

> But another question came into my mind: do you need those annotation copies to be **_actually annotations_**? Would it be ok to just insert them as pixmap images instead? You could create annotation pixmaps on your above pages 2 and 3 (with transparency of course) and insert the pixmaps in page 1.

You cannot create annotations with PyMuPDF in this arbitrarity - you have realized this already. So this way out is blocked.
A feasible approach would need to do a low-level copy of the annotation definitions.

For each annotation to copy over:
* determine all xrefs it references, look at this example. Other annotation types may have more / other references.
```python
<<
  /Type /Annot
  /Subtype /Polygon
  /AP <<
    /N 38 0 R  # <== appearance object: must be completely copied over too
  >>
  /C [ 0 0 1 ]
  /P 4 0 R  # <== xref of the page: replace by target page
  /F 4
  /Vertices [ 63.098249 397.28208 138.09825 432.28208 138.09825
      397.28208 63.098249 432.28208 ]
  /NM (jorj-A8)  # optional, better ignore this
  /Rect [ 62.79825 396.9821 138.39826 432.58207 ]
  /BS <<
    /D [ 2 ]
    /W .3
    /S /D
  >>
  /IC [ 1 1 0 ]
  /LE [ /Diamond /Circle ]
>>
```
* for each of those objects create respective new xrefs in the target PDF, extract (and modify) the definition sources and then insert all of them in the target PDF.
* Modify the target page's object definition (its `/Annots` array) to include xref pointers to the new annotations.

**Comment by amole-arup** [comment]
Hi Jorj, thank you very much for your clear response. How do I get access to the Postscript you are showing? Can I get that using PyMuPDF or should I go in with some other tool (GhostScript or similar)?

**Comment by JorjMcKie** [comment]
You can freely access all PDF source identified by an xref via `doc.xref_object(annot.xref, compressed=True/False)`.
"compressed" False gives you the syntax like above. True will deliver a string without newliners and unnecessary spaces.
Likewise you can access parts of that source via `doc.xref_get_key(annot.xref, "Subtype")`, which will return `("name", "/Polygon")` in the above case. This latter feature also supports dictionary hierarchies: `doc.xref_get_key(annot.xref, "AP/N")` will give you the xref string of the normal appearance dictionary in the form `("xref", "38 0 R")`.

Vice versa it is possible to use PyMuPDF like an editor for these things:
* `doc.xref_update(annot.xref, modified_source)` using some modified postscript version of above
* `doc.xref_set_key(annot.xref, "C", "[1 0 0]")` which in this case will change the stroke color from blue to red. 

> All this also works for the PDF trailer: just use -1 instead of a positive xref. A unique chance to radically spoil your PDF 😆

**Comment by JorjMcKie** [comment]
sorry: `doc.update_object()` not `doc.xref_update()`!

**Comment by amole-arup** [comment]
Thank you very much. 

Great work, by the way. We previously used PyMuPDF to help pull data from PDF piling records that did not have internal structure - we used search to find the section with the data (based on record headings, then used the location to pull in the form data within a rectangular area. Finally we highlighted the text that we extracted as a record that it had been extracted. It saved some poor graduate engineer from having to do lots of painful data entry (with associated errors, probably).
